proc/put/put.go

```diff
@@ -81,116 +84,279 @@ func (p *Proc) eval(in *zng.Record) ([]zng.Value, error) {
+			return nil, err
+		}
+		return b.Bytes(), nil
+	default:
+		// top-level op should be root or record
+		panic(fmt.Sprintf("put: unexpected step %v", s.op))
 	}
-	var tail []int
-	for k, cl := range p.clauses {
-		typ := vals[k].Type
-		clauseTypes[k] = clauseType{typ, zng.IsContainerType(typ)}
-		name := cl.LHS.String()
-		col := zng.Column{Name: name, Type: typ}
-		position, hasCol := inType.ColumnOfField(name)
-		if hasCol {
-			nreplace++
-			replace[position] = k
-			cols[position] = col
-		} else {
-			tail = append(tail, k)
-			cols = append(cols, col)
+}
+
+func (s step) buildRecord(in zcode.Bytes, b *zcode.Builder, vals []zng.Value) error {
+	ig := newGetter(in)
+
+	for _, step := range s.record {
+		switch step.op {
+		case fromInput:
+			bytes, err := ig.nth(step.index)
+			if err != nil {
+				return err
+			}
+			if step.container {
+				b.AppendContainer(bytes)
+			} else {
+				b.AppendPrimitive(bytes)
+			}
+		case fromClause:
+			if step.container {
+				b.AppendContainer(vals[step.index].Bytes)
+			} else {
+				b.AppendPrimitive(vals[step.index].Bytes)
+			}
+		case record:
+			b.BeginContainer()
+			bytes, err := in, error(nil)
+			if step.index >= 0 {
+				bytes, err = ig.nth(step.index)
+				if err != nil {
+					return err
+				}
+			}
+			err = step.buildRecord(bytes, b, vals)
+			if err != nil {
+				return err
+			}
+			b.EndContainer()
+		}
+	}
+	return nil
+}
+
+// A getter provides random access to values in a zcode container
+// using zcode.Iter. It uses a cursor to avoid quadratic re-seeks for
+// the common case where values are fetched sequentially.
+type getter struct {
+	cursor    int
+	container zcode.Bytes
+	iter      zcode.Iter
+}
+
+func newGetter(cont zcode.Bytes) getter {
+	return getter{
+		cursor:    -1,
+		container: cont,
+		iter:      cont.Iter(),
+	}
+}
+func (ig *getter) nth(n int) (zcode.Bytes, error) {
+	if n < ig.cursor {
+		ig.iter = ig.container.Iter()
+	}
+	for !ig.iter.Done() {
+		zv, _, err := ig.iter.Next()
+		if err != nil {
+			return nil, err
+		}
+		ig.cursor++
+		if ig.cursor == n {
+			return zv, nil
+		}
+	}
+	return nil, fmt.Errorf("getter.nth: array index %d out of bounds", n)
+}
+
+func findOverwriteClause(path field.Static, clauses []expr.Assignment) (int, field.Static, bool) {
+	for i, cand := range clauses {
+		if path.Equal(cand.LHS) || path.IsParent(cand.LHS) {
+			return i, cand.LHS, true
+		}
+	}
+	return -1, nil, false
+}
+
+func (p *Proc) deriveRule(parentPath field.Static, inType *zng.TypeRecord, vals []zng.Value) (step, *zng.TypeRecord, error) {
+	// special case: assign to root (put .=x)
+	if p.clauses[0].LHS.IsRoot() {
+		recVal, ok := vals[0].Type.(*zng.TypeRecord)
+		if !ok {
+			return step{}, nil, fmt.Errorf("put .=x: cannot put a non-record to .")
+		}
+		typ, err := p.pctx.TypeContext.LookupTypeRecord(recVal.Columns)
+		return step{op: root, index: 0}, typ, err
+	}
+	return p.deriveRecordRule(parentPath, inType.Columns, vals)
```

Comment:
I'd get rid of the first argument to this method since it's always `field.NewRoot()`.
```suggestion
func (p *Proc) deriveRule(inType *zng.TypeRecord, vals []zng.Value) (step, *zng.TypeRecord, error) {
	// special case: assign to root (put .=x)
	if p.clauses[0].LHS.IsRoot() {
		recVal, ok := vals[0].Type.(*zng.TypeRecord)
		if !ok {
			return step{}, nil, fmt.Errorf("put .=x: cannot put a non-record to .")
		}
		typ, err := p.pctx.TypeContext.LookupTypeRecord(recVal.Columns)
		return step{op: root, index: 0}, typ, err
	}
	return p.deriveRecordRule(field.NewRoot(), inType.Columns, vals)
```

Reply:
good catch! this was a relic of some earlier step.